README.md
# faviator
> A simple easy favicon generator.

[![Build Status](https://travis-ci.org/faviator/faviator.svg?branch=master)](https://travis-ci.org/faviator/faviator)
[![codecov](https://codecov.io/gh/faviator/faviator/branch/master/graph/badge.svg)](https://codecov.io/gh/faviator/faviator)

![Logo](favicon.png)

## Playground
Manually changing the config and generate the favicon to see the result might be time consuming. A faviator playground is created to allow us to tweak the config and view the result instantly.

[Click here to go to the playground.](https://www.faviator.xyz/playground)

## Install

CLI:
```
npm install -g faviator
```

Programatic API:
```
npm install --save-dev faviator
```

## How to use (CLI)

```
> faviator -h
    Usage: faviator [options]

  A simple easy favicon generator.


  Options:

    -V, --version                   output the version number
    -s, --size <n>                  Width and height of the favicon
    -t, --text <value>              Text in the favicon
    --dx <n>                        Move text horizontally
    --dy <n>                        Move text vertically
    --font-size <n>                 Font size of the text
    -f, --font-family <value>       Font family; please choose from Google Fonts
    --font-weight <value>           Font weight; please choose from Google Fonts
    --font-color <value>            Color name/hex/rgb
    -B, --background-color <value>  Background color of favicon
    --border-width <n>              Width of the border
    -b, --border-color <value>      Color of the border
    -R, --border-radius <n>         Short hand to set rx and ry
    --rx <n>                        x-axis border radius
    --ry <n>                        y-axis border radius
    -c, --config <path>             use a config file to draw
    -o, --output <path>             If not specified, svg will be printed to stdout. You can use .svg/.jpeg/.jpg/.png extensions.
    -h, --help                      output usage information
```

--------

By default, faviator will print the xml of the svg to stdout.

Example:
```
> faviator --size '160' --text 'Fav' --dy '19.5' --font-size '32' --font-family 'Roboto' --font-color 'skyblue' --font-weight '700' --background-color 'navy' --border-width '11.5' --border-color 'skyblue' --border-radius '18.5'
<?xml version="1.0" standalone="no"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
<svg xmlns="http://www.w3.org/2000/svg" width="160" height="160" viewBox="0 0 100 100">
  <defs>
    <style type="text/css">
      @import url('https://fonts.googleapis.com/css?family=Roboto:700');
    </style>
  </defs>

<rect x="5.75"
  y="5.75"
  rx="18.5"
  ry="18.5"
  width="88.5"
  height="88.5"
  fill="navy"></rect>

<text x="50"
  y="50"
  dx="0"
  dy="19.5"
  width="160"
  height="160"
  fill="skyblue"
  font-size="32"
  font-family="Roboto"
  text-anchor="middle"
  alignment-baseline="central">Fav</text>

<rect x="5.75"
  y="5.75"
  rx="18.5"
  ry="18.5"
  width="88.5"
  height="88.5"
  fill-opacity="0"
  stroke="skyblue"
  stroke-width="11.5"></rect>
</svg>
```

![0.svg](example/0.svg)

-----

You can export the svg/jpeg/png to a file with the `-o` flag.
```
> faviator --size '160' \
           --text 'f' \
           --dx '-5' \
           --dy '0' \
           --font-size '55' \
           --font-family 'Tangerine' \
           --font-color '#75b7ff' \
           --font-weight '700' \
           --background-color 'rgb(37, 86, 209)' \
           --border-width '3.5' \
           --border-color '#75b7ff' \
           --border-radius '0' \
           -o example/1.png
```

1.png:

![1.png](example/1.png)

-------

You should also save your config to a json file so that you can reproduce your design.

```
> faviator -c path/to/config.json \
           -o example/2.png
```

2.png:

![2.png](example/2.png)

## How to use? (Programatic API)

Faviator provides three functions that generate the favicon as svg, jpeg or png. The API of the three functions are identical. They all return a promise that resolves to a `Buffer` that represents the favicon.

### config
The config object takes in the following keys and generate the favicon correspondingly. Most values are self-explantory.

The following values are the default values if it is not defined. This defines the faviator's logo.
```javascript
const config = {
  size: 16,            // the width and height of the generated image (in px)
  text: '',
  dx: 0,               // move the text from the 'center' horizontally
  dy: 0,               // move the text from the 'center' vertically
  fontSize: 0,
  fontFamily: '',      // a font from Google Font
  fontColor: '',
  fontWeight: '',      // the weight of the font from Google Font
  backgroundColor: '',
  borderWidth: 0,
  borderColor: '',
  borderRadius: 0,     // rx and ry will be set to this if defined
  rx: 0,               // x-axis radius of the favicon
  ry: 0,               // y-axis radius of the favicon
};
```

### faviator(config) / faviator.svg(config)

Example:
```javascript
const faviator = require('faviator');

faviator.svg(config)
  .then(buffer => buffer.toString())
  .then(svg => console.log(svg));

// output: the xml of the svg
```

### faviator.jpeg(config) / faviator.jpg(config)

Example:
```
const fs = require('fs');
const faviator = require('faviator');

faviator.jpeg(config).then(buffer => fs.writeFileSync('favicon.jpg', buffer));

// or

faviator.jpg(config).then(buffer => fs.writeFileSync('favicon.jpg', buffer));
```

### faviator.png(config)

Example:
```
const fs = require('fs');
const faviator = require('faviator');

faviator.png(config).then(buffer => fs.writeFileSync('favicon.png', buffer));
```

## How to use the icon generated?

After generating the icon, you could simply add the following line in your `<head>` tag.

PNG:
```html
<link rel="icon" href="favicon.png" type="image/png">
```

JPG:
```html
<link rel="icon" href="favicon.jpg" type="image/jpg">
```

## Author
Jason Yu <me@ycmjason.com>


CHANGELOG.md
# Changelog

## 1.5.0
- fix default export issue with createSvgFavicon

## 1.4.0
- update dependencies

## 1.3.2
- replace usage of `new Buffer` to `Buffer.from`

## 1.3.1
- update @favicon/create-svg-icon for font-weight support

## 1.3.0
- Add support for font weight

## 1.2.0

### Feature
- Dropped [convert-svg](https://github.com/NotNinja/convert-svg) and used [@ycm.jason/svg-to-img](https://github.com/ycmjason/svg-to-img) for better image quality and prefetching of resources.

### Chore
- Bump deps

## 1.1.3

### Chore
- Bump deps


## 1.1.2

### Chore
- Updated release pipeline
- `npm audit fix`
- switch to [commit-lint](http://marionebl.github.io/commitlint/#/)


test/index.test.js
const assert = require('assert');
const proxyquire = require('proxyquire');
const sinon = require('sinon');

const createSvgFavicon = require('@faviator/create-svg-favicon');

describe('faviator [index.js]', function() {
  this.timeout(5000);
  beforeEach(function() {
    this.stubs = {
      '@ycm.jason/svg-to-img': {
        jpeg: sinon.stub().resolves('some-jpeg'),
        png: sinon.stub().resolves('some-png'),
      },
    };
    this.faviator = proxyquire('../lib/index', this.stubs);
  });

  it('should call .svg', function(next) {
    sinon.stub(this.faviator, 'svg').resolves('hello');
    this.faviator().then(v => {
      assert.equal(v, 'hello');
      assert(this.faviator.svg.called);
    }).then(next);
  });

  describe('.svg', function() {
    it('should wrap createSvgFavicon() in a Promise', function(next) {
      this.faviator.svg().then(svg => {
        assert.equal(svg.toString(), createSvgFavicon());
      }).then(() => next(), next);
    });
  });

  describe('.jpg', function() {

    it('should have alias .jpeg', function() {
      assert.equal(this.faviator.jpeg, this.faviator.jpg);
    });

    it('should call svgToImg.jpeg', function(next) {
      this.faviator.jpg().then(v => {
        assert.equal(v, 'some-jpeg');
        assert(this.stubs['@ycm.jason/svg-to-img'].jpeg.called);
        next();
      });
    });

  });

  describe('.png', function() {

    it('should call svgToImg.jpeg', function(next) {
      this.faviator.png().then(v => {
        assert.equal(v, 'some-png');
        assert(this.stubs['@ycm.jason/svg-to-img'].png.called);
        next();
      });
    });

  });

  describe('.faviatorIconConfig', function() {

    it('should not be undefined', function() {
      assert(typeof this.faviator.faviatorIconConfig === 'object');
      assert(this.faviator.faviatorIconConfig !== null);
    });

  });
});


lib/__tests__/index.test.js
const faviator = require('../index');
const createSvgFavicon = require('@faviator/create-svg-favicon').default;
const svgToImgMock = require('@ycm.jason/svg-to-img');

jest.mock('@ycm.jason/svg-to-img');

describe('faviator [index.js]', () => {
  it('should call .svg', async () => {
    jest.spyOn(faviator, 'svg').mockResolvedValue('hello');
    const v = await faviator();
    expect(v).toEqual('hello');
  });

  describe('.svg', () => {
    it('should wrap createSvgFavicon() in a Promise', async () => {
      const svg = await faviator.svg();
      expect(svg.toString()).toBe(createSvgFavicon());
    });
  });

  describe('.jpg', () => {
    it('should have alias .jpeg', () => {
      expect(faviator.jpeg).toBe(faviator.jpg);
    });

    it('should call svgToImg.jpeg', async () => {
      const expectJpg = 'some-jpeg';
      svgToImgMock.jpeg.mockResolvedValue(expectJpg);
      const jpg = await faviator.jpg();
      expect(jpg).toBe(expectJpg);
    });
  });

  describe('.png', () => {
    it('should call svgToImg.jpeg', async () => {
      const expectPng = 'some-png';
      svgToImgMock.png.mockResolvedValue(expectPng);
      const png = await faviator.png();
      expect(png).toEqual(expectPng);
    });
  });

  describe('.faviatorIconConfig', () => {
    it('should not be undefined', () => {
      expect(faviator.faviatorIconConfig).toBeDefined();
      expect(typeof faviator.faviatorIconConfig).toBe('object');
      expect(faviator.faviatorIconConfig).not.toBeNull();
    });
  });
});


test/lib/createSvgFavicon.test.js
const assert = require('assert');

const createSvgFavicon = require('../../lib/createSvgFavicon');
const removeNewLines = (s) => s.replace(/(\r|\n)/g, '');

describe('createSvgFavicon', function() {
  it('should generate the doctype tag', function() {
    assert((/<!doctype svg/i).test(createSvgFavicon()));
  });

  it('should generate a svg tag with width and height as specified in config.size', function() {
    const svg = removeNewLines(createSvgFavicon({ size: 100 }));
    assert.equal(svg.match(/<svg [^>]*?height="([^"]*)"/i)[1], '100');
    assert.equal(svg.match(/<svg [^>]*?width="([^"]*)"/i)[1], '100');
  });

  it('should generate a rect tag with rx, ry, fill', function() {
    const rx = '50%';
    const ry = '50%';
    const backgroundColor = '#03f53a';
    const svg = removeNewLines(createSvgFavicon({
      rx,
      ry,
      backgroundColor,
    }));
    assert.equal(svg.match(new RegExp(`<rect [^>]*?rx="([^"]*)"`, 'i'))[1], rx);
    assert.equal(svg.match(new RegExp(`<rect [^>]*?ry="([^"]*)"`, 'i'))[1], ry);
    assert.equal(svg.match(new RegExp(`<rect [^>]*?fill="([^"]*)"`, 'i'))[1], backgroundColor);
  });

  it('should generate a rect tag with rx, ry with value borderRadius', function() {
    const borderRadius = '50%';
    const svg = removeNewLines(createSvgFavicon({
      borderRadius,
    }));
    assert.equal(svg.match(new RegExp(`<rect [^>]*?rx="([^"]*)"`, 'i'))[1], borderRadius);
    assert.equal(svg.match(new RegExp(`<rect [^>]*?rx="([^"]*)"`, 'i'))[1], borderRadius);
  });

  it('should generate a text tag with content text, fill, font-family, font-size', function() {
    const text = 'love';
    const fontColor = 'red';
    const fontFamily = 'Open Sans';
    const fontSize = 65;
    const svg = removeNewLines(createSvgFavicon({
      text,
      fontColor,
      fontFamily,
      fontSize,
    }));
    assert.equal(svg.match(new RegExp(`<text.*?>([^<]*)</text>`, 'i'))[1], text);
    assert.equal(svg.match(new RegExp(`<text [^>]*?fill="([^"]*)"`, 'i'))[1], fontColor);
    assert.equal(svg.match(new RegExp(`<text [^>]*?font-family="([^"]*)"`, 'i'))[1], fontFamily);
    assert.equal(svg.match(new RegExp(`<text [^>]*?font-size="([^"]*)"`, 'i'))[1], fontSize);
  });

  it('should generate a text tag with at center', function() {
    const svg = removeNewLines(createSvgFavicon());
    assert.equal(svg.match(new RegExp(`<text [^>]*?x="([^"]*)"`, 'i'))[1], '50');
    assert.equal(svg.match(new RegExp(`<text [^>]*?y="([^"]*)"`, 'i'))[1], '50');
    assert.equal(svg.match(new RegExp(`<text [^>]*?text-anchor="([^"]*)"`, 'i'))[1], 'middle');
    assert.equal(svg.match(new RegExp(`<text [^>]*?alignment-baseline="([^"]*)"`, 'i'))[1], 'central');
  });
});


lib/createSvgFavicon.js
module.exports = ({
  size = 16,
  text = 'F',
  dx = 0,
  dy = 0,
  fontSize = 80,
  fontFamily = 'Dancing Script',
  fontColor = 'white',
  backgroundColor = 'rgb(219, 59, 211)',
  borderWidth = 5,
  borderColor = '#0D1423',
  borderRadius,
  rx = '5%',
  ry = '5%',
} = {}) => `<?xml version="1.0" standalone="no"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
<svg xmlns="http://www.w3.org/2000/svg" width="${size}" height="${size}" viewBox="0 0 100 100">
  <defs>
    <style type="text/css">
      @import url('https://fonts.googleapis.com/css?family=${fontFamily.replace(/ /g, '+')}');
    </style>
  </defs>

<rect x="${parseFloat(borderWidth) / 2}"
  y="${parseFloat(borderWidth) / 2}"
  rx="${borderRadius || rx}"
  ry="${borderRadius || ry}"
  width="${100 - parseFloat(borderWidth)}"
  height="${100 - parseFloat(borderWidth)}"
  fill="${backgroundColor}"></rect>

<text x="50"
  y="50"
  dx="${dx}"
  dy="${dy}"
  width="${size}"
  height="${size}"
  fill="${fontColor}"
  font-size="${fontSize}"
  font-family="${fontFamily}"
  text-anchor="middle"
  alignment-baseline="central">${text}</text>

<rect x="${parseFloat(borderWidth) / 2}"
  y="${parseFloat(borderWidth) / 2}"
  rx="${borderRadius || rx}"
  ry="${borderRadius || ry}"
  width="${100 - parseFloat(borderWidth)}"
  height="${100 - parseFloat(borderWidth)}"
  fill-opacity="0"
  stroke="${borderColor}"
  stroke-width="${borderWidth}"></rect>
</svg>`;
